File: src/Functional/Clojure/example2.clj
(ns example2)

(defprotocol DatabaseFactory
  (connect [this])
  (query [this]))

(defrecord PostgresFactory []
  DatabaseFactory
  (connect [_] (println "Connecting to PostgreSQL"))
  (query [_] (println "Querying PostgreSQL")))

(defrecord MySQLFactory []
  DatabaseFactory
  (connect [_] (println "Connecting to MySQL"))
  (query [_] (println "Querying MySQL")))

(defn use-database [factory]
  (connect factory)
  (query factory))

;; Usage
(use-database (->PostgresFactory))
(use-database (->MySQLFactory))

File: src/Functional/Erlang/example2.erl
-module(example2).
-export([connect_db/1, query_db/1]).

% Abstract Factory
connect_db(postgresql) -> connect_postgresql();
connect_db(mysql) -> connect_mysql().

query_db(postgresql) -> query_postgresql();
query_db(mysql) -> query_mysql().

% Concrete Implementations
connect_postgresql() -> io:format("Connecting to PostgreSQL~n").
query_postgresql() -> io:format("Querying PostgreSQL~n").

connect_mysql() -> io:format("Connecting to MySQL~n").
query_mysql() -> io:format("Querying MySQL~n").

% Usage
main() ->
    connect_db(postgresql),
    query_db(postgresql),
    connect_db(mysql),
    query_db(mysql).
